homeassistant/components/almond/__init__.py

```diff
@@ -210,21 +210,13 @@ def __init__(self, api: WebAlmondAPI):
         response = await self.api.async_converse_text(text)
 
         buffer = ""
+        extra_data = []
         for message in response["messages"]:
             if message["type"] == "text":
                 buffer += "\n" + message["text"]
-            elif message["type"] == "picture":
-                buffer += "\n Picture: " + message["url"]
-            elif message["type"] == "rdl":
-                buffer += (
-                    "\n Link: "
-                    + message["rdl"]["displayTitle"]
-                    + " "
-                    + message["rdl"]["webCallback"]
-                )
-            elif message["type"] == "choice":
-                buffer += "\n Choice: " + message["title"]
+            else:
+                extra_data.append(message)
 
         intent_result = intent.IntentResponse()
-        intent_result.async_set_speech(buffer.strip())
+        intent_result.async_set_speech(buffer.strip(), extra_data=extra_data)
```

Comment:
@gcompax is right. We can't just throw things in extra data. That's bypassing the Home Assistant abstraction.

The frontend should only work with the Home Assistant abstraction which is `IntentResponse`. Intent Response should be able to support the different options that Almond supports.

Reply:
I agree completely, it should be abstracted, but still we need to split out text from links and images. For a pure speech frontend it doesn't make sense to speak out the url of a image. For choice, yes, it makes sense, but if there is a display you would also want to show it.

So let's come up with a good abstraction.

Reply:
By the way, Almond has multiple kinds of links/buttons, that probably need to be treated differently.

`type=rdl` (rich deep link) are external links in the form of cards, with title, optional description and optional picture. You'll definitely want to speak the title and maybe the description of these, otherwise things like "show me the news" which return RDL results will say nothing.
`type=link` are internal links inside Almond website, and they're used to link to things like configuring new skills, that cannot be done inside the chat. You'll probably want to show these as links/buttons (editing the URL to point to Almond), or ignore them. You shouldn't speak them though, they're not useful in dialog.
`type=choice` are contextual multiple choice, used for example to choose "which light do you want to turn on?". You have to speak the title of these, otherwise the user doesn't know what are the available choices. You don't have to display these necessarily as buttons - the user can reply in text.
`type=button` are suggestions (eg. "Show more results..."). You should at least speak the title of these. In some contexts, it's good to have them as clickable buttons, because they are used to override the voice/NLP when they fail.
`type=result` are open-ended interactive card-like results - they are a catch-all for future extensions to the protocol, and are designed to display videos, maps, etc. There is a "fallback" property that you should use in voice.

On top of that, Almond expects that in certain contexts "picker buttons" will show up. For example, if I say "post a picture on twitter", Almond says "Upload the picture now.", and the UI should show a button that pops up a file chooser. This is indicated by the "askSpecial" field of the response. Most "askSpecial" values are just enhancements; "picture" / "file" are the one case where a file chooser is necessary.

(Documentation:
https://almond.stanford.edu/doc/almond-dialog-api-reference.md

Also, here is the code that speaks things in Almond natively:
https://github.com/stanford-oval/almond-server/blob/master/service/assistant.js
look for `say` inside `MainConversation`)